transaction/core/src/validation/validate.rs

```diff
@@ -1375,4 +1393,50 @@ mod tests {
             )
         }
     }
+
+    // Test that input rules validation is working
+    #[test]
+    fn test_input_rules_validation() {
+        let block_version = BlockVersion::THREE;
+
+        for _ in 0..3 {
```

Comment:
yeah that's wierd -- there should be some randomness in here somewhere that would make this more reasonable. but it seems that there isn't. i'll remove the loop i guess

Reply:
This is a pretty good question.

It is obviously not very good to reveal the amount of a TxOut given that our general goal is not to reveal that.

However:
* In any sort of "proposal to trade", the counterparty who accepts the trade necessarily has some insight into what they traded, and also, what outputs were created by the Tx that they submitted. So as a practical matter I don't know how we can avoid this.
* Assuming that is unavoidable, this scheme still doesn't reveal the public address of who is receiving the amount.
* This scheme also doesn't reveal who submitted the offer to trade (the originator) to the counter-party, or at least, it doesn't necessarily do so.
* The TxOut would only be "penalized for spending" if it was obvious when someone was spending it. But by design our network makes it very difficult to figure out if any TxOut has been spent. No one can link the key image back to the TxOut unless they have the spend private key.

So I think, the "penalized for spending" thing is not real, and this is likely the best we can do for now. If we had MobZero then we could also avoid revealing the input rings of the true input -- that's the most feasible way that I can see that we can improve this.

You might imagine that there's a system where like, instead of sending your trade proposal to a network of counterparties, you could propose it to some kind of zero-knowledge smart contract. But right now I'm not sure I believe that that is real. I don't believe right now that e.g. you can have a DEX on chain that doesn't expose enough state to be able to infer what it's treasury is, and therefore also infer individual changes in its treasury value. I don't think an off-chain SGX thing works either because SGX doesn't really allow for durable state, all the state in an enclave has to be designed to be discardable. But the DEX private keys are obviously not discardable.

So right now the only thing that makes sense to me is like, having a network which makes it difficult to infer who submitted a particular message, and accepting that messages will reveal the amounts of some TxOut's. Happy to be wrong.